File: src/main/java/com/ploy/bubble_server_v3/domain/machine/service/WashingMachineService.java
package com.ploy.bubble_server_v3.domain.machine.service;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.ploy.bubble_server_v3.domain.machine.domain.WashingMachine;
import com.ploy.bubble_server_v3.domain.machine.exception.WashingMachineDataFetchException;
import com.ploy.bubble_server_v3.domain.machine.infra.dto.response.WashingMachineResponse;
import com.ploy.bubble_server_v3.domain.machine.infra.feign.WashingMachineFeign;
import com.ploy.bubble_server_v3.domain.machine.util.DeviceProcessor;
import com.ploy.bubble_server_v3.domain.machine.util.HeaderProvider;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.redis.core.RedisTemplate;
import org.springframework.stereotype.Service;

import java.util.*;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class WashingMachineService {

    private final WashingMachineFeign washingMachineFeign;
    private final HeaderProvider headerProvider;
    private final DeviceProcessor deviceProcessor;
    private final ObjectMapper objectMapper;
    private final RedisTemplate<String, Object> redisTemplate;

    private static final String REDIS_KEY_PREFIX = "washingMachineStatus:";

    private static final List<String> SORT_ORDER = Arrays.asList(
            "세탁기1", "세탁기2", "세탁기3", "세탁기4", "건조기1", "건조기2"
    );

    public List<WashingMachine> getAllWashingMachines() {
        WashingMachineResponse response = fetchWashingMachineData();

        List<WashingMachine> result = new ArrayList<>();

        if (response != null && response.result() != null && response.result().devices() != null) {
            response.result().devices().forEach(device ->
                    deviceProcessor.processAllDevices(device, result));
        }

        return result;
    }

    public List<WashingMachine> getWashingMachinesByRoom(String roomId) {
        WashingMachineResponse response = fetchWashingMachineData();

        List<WashingMachine> result = new ArrayList<>();

        if (response != null && response.result() != null && response.result().devices() != null) {
            List<WashingMachine> finalResult = result;
            response.result().devices().forEach(device ->
                    deviceProcessor.processDeviceByRoom(device, finalResult, roomId));
        }

        result = sortMachines(result);

        return result;
    }

    public void saveWashingMachines(List<WashingMachine> washingMachineList) {
        String redisKey = REDIS_KEY_PREFIX + "all";
        try {
            String json = objectMapper.writeValueAsString(washingMachineList);
            redisTemplate.opsForValue().set(redisKey, json);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("JSON 직렬화 실패", e);
        }
    }

    public List<WashingMachine> getAllWashingMachinesFromRedis() {
        String redisKey = REDIS_KEY_PREFIX + "all";
        Object value = redisTemplate.opsForValue().get(redisKey);

        log.info("레디스 조회 성공 : " + value);

        try {

            if (value instanceof String json) {
                List<WashingMachine> list = objectMapper.readValue(json, new TypeReference<>() {});
                return list;
            }
        } catch (JsonProcessingException e) {
            throw new RuntimeException("JSON 역직렬화 실패", e);
        }
        return null;
    }


    private WashingMachineResponse fetchWashingMachineData() {
        try {
            return washingMachineFeign.getWashingMachineData(headerProvider.getRequestHeaders());
        } catch (Exception e) {
            throw new WashingMachineDataFetchException();
        }
    }

    private List<WashingMachine> sortMachines(List<WashingMachine> machines) {
        return machines.stream()
                .sorted(Comparator.comparing(machine -> {
                    int index = SORT_ORDER.indexOf(machine.name());
                    return index != -1 ? index : SORT_ORDER.size();
                }))
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/com/ploy/bubble_server_v2/domain/washingMachine/service/WashingMachineService.kt
package com.ploy.bubble_server_v2.domain.washingMachine.service


import com.laundering.laundering_server.common.exception.BusinessException

@lombok.extern.slf4j.Slf4j
@org.springframework.stereotype.Service
@RequiredArgsConstructor
class WashingMachineService {
    @Autowired
    private val userRepository: UserRepository? = null

    @Autowired
    private val restTemplate: RestTemplate? = null

    @Autowired
    private val objectMapper: com.fasterxml.jackson.databind.ObjectMapper? = null

    fun getStatus(id: Long?): List<WashingMachineResponse> {
        try {
            // 주어진 id로 사용자 정보를 조회
            val user: User = userRepository.findById(id)
                .orElseThrow { BusinessException(ErrorCode.UNAUTHORIZED) }

            // 세탁실 정보 가져오기
            val washingRoom: String = user.getWashingRoom()

            // URL을 구성하고 HTTP GET 요청 보내기
            val url = String.format("https://build-bubble-proxy-server.vercel.app/home/%s", washingRoom)
            val response: String = restTemplate.getForObject<String>(url, String::class.java)

            // JSON 응답을 WashingMachineResponse 리스트로 변환하기
            return objectMapper!!.readValue<List<WashingMachineResponse>>(
                response,
                object : com.fasterxml.jackson.core.type.TypeReference<List<WashingMachineResponse?>?>() {})
        } catch (e: IOException) {
            // JSON 처리 중 오류가 발생한 경우 또는 IO 오류가 발생한 경우 처리
            throw BusinessException(ErrorCode.BAD_REQUEST)
        }
    }
}


